conans/model/settings.py

```diff
@@ -225,6 +229,13 @@ def validate(self):
             child = self._data[field]
             child.validate()
 
+    def remove_undefined(self):
+        for field, data in list(self._data.items()):
+            if data.value is None:
+                self._data.pop(field)
+            else:
+                data.remove_undefined()
```

Comment:
As settings model is a complexity nightmare, would be useful to know what is "data" here to follow the method. Also a docstring about what is doing the method. At first sight I wonder, could it be introducing a bug by removing some setting without a value instead of raise an error of not declared value for a setting? Oviously I don't understand it.